packages/graph/hash_graph/lib/graph/src/store/postgres/query/condition.rs

```diff
@@ -0,0 +1,328 @@
+use std::{fmt, fmt::Write};
+
+use postgres_types::ToSql;
+
+use crate::store::{
+    postgres::query::{
+        database::{Column, Table, TableAlias},
+        Path, PostgresQueryRecord, Transpile,
+    },
+    query::{Filter, FilterValue, Parameter},
+};
+
+/// A [`Filter`] compiled to postgres.
+pub enum Condition<'q> {
+    All(Vec<Self>),
+    Any(Vec<Self>),
+    Not(Box<Self>),
+    Equal(Option<ConditionValue<'q>>, Option<ConditionValue<'q>>),
+    NotEqual(Option<ConditionValue<'q>>, Option<ConditionValue<'q>>),
+}
+
+impl<'q> Condition<'q> {
+    /// Compiles a [`Filter`] to a `Condition`.
+    ///
+    /// In addition to the provided [`Filter`], a list of parameters needs to be passed, which will
+    /// be populated from [`FilterValue::Parameter`]. Also, [`TableAlias`] is used to populate
+    /// [`Table::alias`] inside of [`ConditionValue::Column`].
+    pub fn from_filter<'f: 'q, T: PostgresQueryRecord<'q>>(
+        filter: &'f Filter<'q, T>,
+        parameters: &mut Vec<&'f dyn ToSql>,
+        alias: Option<TableAlias>,
+    ) -> Self {
+        match filter {
+            Filter::All(filters) => Self::All(
+                filters
+                    .iter()
+                    .map(|filter| Self::from_filter(filter, parameters, alias))
+                    .collect(),
+            ),
+            Filter::Any(filters) => Self::Any(
+                filters
+                    .iter()
+                    .map(|filter| Self::from_filter(filter, parameters, alias))
+                    .collect(),
+            ),
+            Filter::Not(filter) => {
+                Self::Not(Box::new(Self::from_filter(filter, parameters, alias)))
+            }
+            Filter::Equal(lhs, rhs) => Self::Equal(
+                lhs.as_ref()
+                    .map(|value| ConditionValue::from_filter_value(value, parameters, alias)),
+                rhs.as_ref()
+                    .map(|value| ConditionValue::from_filter_value(value, parameters, alias)),
+            ),
+            Filter::NotEqual(lhs, rhs) => Self::NotEqual(
+                lhs.as_ref()
+                    .map(|value| ConditionValue::from_filter_value(value, parameters, alias)),
+                rhs.as_ref()
+                    .map(|value| ConditionValue::from_filter_value(value, parameters, alias)),
+            ),
+        }
+    }
+}
+
+impl Transpile for Condition<'_> {
+    fn transpile(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Condition::All(conditions) if conditions.is_empty() => fmt.write_str("true"),
+            Condition::Any(conditions) if conditions.is_empty() => fmt.write_str("false"),
```

Comment:
Nit:
It seems that the uppercase variant of true/false may be better. Don't think this is blocking, though

`The key words TRUE and FALSE are the preferred (SQL-compliant) method for writing Boolean constants in SQL queries` 
from https://www.postgresql.org/docs/current/datatype-boolean.html